cpp/src/arrow/util/checked_cast.h

```diff
@@ -39,11 +40,11 @@ inline OutputType checked_cast(InputType&& value) {
 }
 
 template <class T, class U>
-std::shared_ptr<T> checked_pointer_cast(const std::shared_ptr<U>& r) noexcept {
+std::shared_ptr<T> checked_pointer_cast(std::shared_ptr<U> r) noexcept {
```

Comment:
I can add an rvalue overload of checked_pointer_cast as the standard does, but my understanding is that since we're always taking ownership of `r` then the copy may as well take place in argument initialization, as with `modernize-pass-by-value`